### PR TITLE
Fix to handle successful run with Uint64 overflow for multiple opcodes

### PR DESCRIPTION
**Pull request by silathdiir**
### Description

Fix successful run cases with Uint64 overflow for multiple opcodes.

1. Add `WordByteRangeGadget` to constrain if Word is within the specified byte range.

2. Add `WordByteCapGadget` to constrain if Word is within the specified byte range (implemented by WordByteRangeGadget) and less than a maximum cap (used to replace a WordByteRangeGadget and LtGadget).

3. Fix bus-mapping and zkevm-circuits to handle overflow cases. And add unit-tests for these cases.

TODO: will try to handle memory offset overflow with zero length in another PR (try to rebase for this local PR https://github.com/scroll-tech/zkevm-circuits/pull/393) and related issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1301.

### Rationale

Reference detailed code in `go-etherum` as:

. [BLOCKHASH](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L438)
. [CALLDATALOAD](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L285)
. [CALLDATACOPY](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L306)
. [CODECOPY](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L364)
. [EXTCODECOPY](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L382)
. [JUMPI](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L550)

### Issue Link

Close https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1276

### Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

Add unit-test cases for Uint64 overflow values.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs [hunk 1]
```diff
@@ -1039,3 +1039,125 @@ impl<F: Field> CommonErrorGadget<F> {
...
+    }
+
+    /// Return true if within the range, false if overflow.
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        original: U256,
+    ) -> Result<bool, Error> {
+        debug_assert!(VALID_BYTES < 32);
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:1132]: I think this `debug_assert` and the one in `valid_value` will never be triggered: this type has a single constructor `construct` which already has this assert, so if we manage to create the the variable it means `VALID_BYTES < 32`.  So any time we call `foo.assign()` or `foo.valid_value()` we know for sure that `VALID_BYTES < 32`, because it has been checked at `let foo = WordByteRangeGadget::construct()`.

File: zkevm-circuits/src/evm_circuit/execution/jumpi.rs [hunk 2]
```diff
@@ -47,18 +46,25 @@ impl<F: Field> ExecutionGadget<F> for JumpiGadget<F> {
...
             );
+
+            cb.opcode_lookup_at(dest.valid_value(), OpcodeId::JUMPDEST.expr(), 1.expr());
         });
 
+        cb.require_zero(
+            "JUMPI condition must be 0 if destination is Uint64 overflow",
+            dest.overflow() * not::expr(is_condition_zero.expr()),
+        );
+
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/jumpi.rs:62]: I think this constraint is redundant with `"JUMPI destination must be within range if condition is non-zero"`:
First one is:
- should_jump == 1 -> dest.within_range == 1
Second one is:
- dest.overflow == 1 -> not(is_condition_zero) == 0

And we have:
- should_jump = not(is_condition_zero)
- dest.within_range = not(dest.overflow)

So the second one can be rewritten as:
- dest.within_range == 0 -> should_jump == 0
which is:
- not(dest.within_range == 1) -> not(should_jump == 1)

And this is equivalent to the first one following the contraposition rule https://en.wikipedia.org/wiki/Contraposition

Let me know if I missed something in my reasoning.

File: zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs [hunk 3]
```diff
@@ -80,39 +70,31 @@ impl<F: Field> ExecutionGadget<F> for ErrorInvalidJumpGadget<F> {
         });
 
         // Look up bytecode length
-        let code_len = cb.query_cell();
         cb.bytecode_length(cb.curr.state.code_hash.expr(), code_len.expr());
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs:73]: Aren't we missing the lookup to the bytecode to verify `value` and `is_code`?

File: zkevm-circuits/src/evm_circuit/execution/extcodecopy.rs [hunk 4]
```diff
@@ -50,17 +50,17 @@ impl<F: Field> ExecutionGadget<F> for ExtcodecopyGadget<F> {
         let external_address =
             from_bytes::expr(&external_address_word.cells[..N_BYTES_ACCOUNT_ADDRESS]);
 
-        let memory_offset = cb.query_cell_phase2();
-        let data_offset = cb.query_word_rlc();
+        let code_size = cb.query_cell();
+
         let memory_length = cb.query_word_rlc();
+        let memory_offset = cb.query_cell_phase2();
+        let code_offset = WordByteCapGadget::construct(cb, code_size.expr());
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/extcodecopy.rs:57]: We're using `WordByteCapGadget` for `code_offset`, but I don't see any place where we check `code_offset.overflow` or `code_offset.within_range`.  So the internal `IsZero` gadget seems to be unused?

File: zkevm-circuits/src/evm_circuit/execution/codecopy.rs [hunk 5]
```diff
@@ -49,14 +49,15 @@ impl<F: Field> ExecutionGadget<F> for CodeCopyGadget<F> {
...
         let opcode = cb.query_cell();
 
-        // Query elements to be popped from the stack.
-        let dst_memory_offset = cb.query_cell_phase2();
-        let code_offset = cb.query_word_rlc();
+        let code_size = cb.query_cell();
+
         let size = cb.query_word_rlc();
+        let dst_memory_offset = cb.query_cell_phase2();
+        let code_offset = WordByteCapGadget::construct(cb, code_size.expr());
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/codecopy.rs:56]: We're using `WordByteCapGadget` for `code_offset`, but I don't see any place where we check `code_offset.overflow` or `code_offset.within_range`.  So the internal `IsZero` gadget seems to be unused?

**Review by ed255 — COMMENTED** [review]
First pass review, so far I've reviewed:
- common_gadget
- circuits: jumpi, extcodecopy, codecopy, error_invalid_jump

File: zkevm-circuits/src/evm_circuit/execution/blockhash.rs [hunk 6]
```diff
@@ -36,38 +35,39 @@ impl<F: Field> ExecutionGadget<F> for BlockHashGadget<F> {
...
             None,
             current_block_number.expr(),
         );
 
-        let block_lt = LtGadget::construct(
-            cb,
-            from_bytes::expr(&block_number.cells),
-            current_block_number.expr(),
-        );
+        let block_number = WordByteCapGadget::construct(cb, current_block_number.expr());
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/blockhash.rs:45]: We're using `WordByteCapGadget` for `block_number`, but I don't see any place where we check `block_number.overflow` or `block_number.within_range`. So the internal `IsZero` gadget seems to be unused?

**Review by ed255 — COMMENTED** [review]
Review completed: In general everything looks correct! My only notes are about a redundant constraint, and a few IsZero gadget that seem to be unused.  I think they should be used: the Lt check is only correct if the value uses the specified number of bytes, and that's what the IsZero gadget inside of WordByteRangeGadget can check.  So probably the logic should be something like: 
```rust
            let src_addr = select::expr(
                code_offset.overflow(), 
                code_size.expr(), 
                select::expr(
                   code_offset.lt_cap(),
                   code_offset.valid_value(),
                   code_size.expr()),
            );
```

**Comment by silathdiir** [comment]
> Review completed ...

Thanks for review. I will try to resolve merge conflicts and review comments. Thanks.

**Review by lispc — APPROVED** [review]

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs (hunk 1 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: Delete `debug_assert!(VALID_BYTES < 32)` in `assign` and `valid_value` functions (only leave it in `construct`) in commit [29cf64c](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/29cf64cc21e305a9248a649f1f8e7fb89917f985).

File: zkevm-circuits/src/evm_circuit/execution/extcodecopy.rs (hunk 4 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/extcodecopy.rs:58]: The `within_range` is invoked in [WordByteCapGadget::construct](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4419f2bdbaca12900df3d577556e28777b4823a3/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L1052) to get valid value in range, then compare with `cap` (code-length for `CODECOPY`). It is used to get the correct memory start.

Since the code-offset is set to `u64::MAX` if overflow in [geth opCodeCopy](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L363). And the memory offset is reset to code-length (if code-offset is greater than code-length) in [getData](https://github.com/ethereum/go-ethereum/blob/66c0c4e5174a0b231727f27c7ad48624189ecc62/core/vm/common.go#L57).

The `lt_cap` is used to set memory start in [this code](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/a3226b4d3c55f61a7cd9ef1258f6b7fc8d5c9d41/zkevm-circuits/src/evm_circuit/execution/codecopy.rs#L85).

BTW, the code-length could not be Uint64 overflow (otherwise raise an error of `ErrGasUintOverflow`).

File: zkevm-circuits/src/evm_circuit/execution/jumpi.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/jumpi.rs]: Delete this redundant constraint in commit [787093f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/787093ffc611c084e83d786ff64245768163aff3).

**Comment by silathdiir** [comment]
> Review completed: In general everything looks correct! My only notes are about a redundant constraint, and a few IsZero gadget that seem to be unused. I think they should be used: the Lt check is only correct if the value uses the specified number of bytes, and that's what the IsZero gadget inside of WordByteRangeGadget can check. So probably the logic should be something like:
> 
> ```rust
>             let src_addr = select::expr(
>                 code_offset.overflow(), 
>                 code_size.expr(), 
>                 select::expr(
>                    code_offset.lt_cap(),
>                    code_offset.valid_value(),
>                    code_size.expr()),
>             );
> ```

Seems that `lt_cap` contains the logic of `within_range` as `with_range && offset < cap` in [WordByteCapGadget::construct](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4419f2bdbaca12900df3d577556e28777b4823a3/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L1052). Could we use `lt_cap` to check directly (as current)?
```
            let src_addr = select::expr(
                code_offset.lt_cap(),
                code_offset.valid_value(),
                code_size.expr(),
            );
```

**Comment by silathdiir** [comment]
> LGTM
> 
> Amazing work @silathdiir
> 
> Also I see that this PR also activates `EXTCODECOPY` and there's a bunch of tests that are not passing, but nice to address them in another PRs.
> 
> e.g. `cargo run --release -- --inspect "extcodehashEmpty_d1(storage)_g0_v0"`
> 

It may be caused by `memory_offset + memory_length` should not be Uint64 overflow. I will check if it could be fixed by local PR https://github.com/scroll-tech/zkevm-circuits/pull/393. Thanks.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs [hunk 7]
```diff
@@ -1039,3 +1039,125 @@ impl<F: Field> CommonErrorGadget<F> {
...
+        not::expr(self.within_range())
+    }
+
+    pub(crate) fn valid_value(&self) -> Expression<F> {
+        debug_assert!(VALID_BYTES < 32);
+
+        from_bytes::expr(&self.original.cells[..VALID_BYTES])
+    }
+
+    pub(crate) fn within_range(&self) -> Expression<F> {
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: Rename `within_range` to `not_overflow` in commit [11f8eb3](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/11f8eb326a88c65be6c234856fca8102a29fbbb0).

File: zkevm-circuits/src/evm_circuit/execution/calldataload.rs [hunk 8]
```diff
@@ -176,62 +210,60 @@ impl<F: Field> ExecutionGadget<F> for CallDataLoadGadget<F> {
...
+            .checked_add(data_offset)
+            .unwrap_or(src_addr_end)
+            .min(src_addr_end);
+
+        let mut calldata_bytes = vec![0u8; N_BYTES_WORD];
+        if offset_within_range {
+            for (i, byte) in calldata_bytes.iter_mut().enumerate() {
+                if call.is_root {
+                    // Fetch from tx call data.
+                    if src_addr.checked_add(i as u64).unwrap() < tx.call_data_length as u64 {
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/calldataload.rs]: Replace `check_add` and `unwrap` with just adding in commit [594553d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/594553d6f19d4f8d6ed77198b7b519bb4297d833).

File: zkevm-circuits/src/evm_circuit/execution/calldataload.rs [hunk 9]
```diff
@@ -176,62 +210,60 @@ impl<F: Field> ExecutionGadget<F> for CallDataLoadGadget<F> {
...
+                        *byte = tx.call_data[src_addr as usize + i];
+                    }
+                } else {
+                    // Fetch from memory.
+                    if src_addr.checked_add(i as u64).unwrap()
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/calldataload.rs]: Replace check_add and unwrap with just adding in commit [594553d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/594553d6f19d4f8d6ed77198b7b519bb4297d833).

File: zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs (hunk 3 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs:73]: It seems that the bytecode is lookup when destination is in valid range in [the code below](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/a3226b4d3c55f61a7cd9ef1258f6b7fc8d5c9d41/zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs#L77).

Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/error_invalid_jump.rs:73]: thanks, you're right!
